### DISCUSSION TITLE
Coinbase Smart Wallet + Paymaster

### DISCUSSION THREAD

**Discussion by Ncookiez** [rainbow-me/rainbowkit#2020] [Q&A]
Hey all - was looking to add a paymaster to abstract some gas for users using our app now that rainbowkit supports the coinbase smart wallet. From coinbase's dev server I was told to follow [this](https://www.smartwallet.dev/guides/paymasters) guide to set it up. Basically using the experimental `useCapabilities` and `useWriteContracts` from wagmi, setting up a proxy to determine transactions to sponsor, etc. This would be a huge undertaking to support and add fallbacks for normal wallets and other networks, and I was hoping this is something rainbowkit could help abstract away.

The ideal integration route I'm envisioning would be something along the lines of passing a paymaster URL and an optional function that returns if a transaction should be sponsored or not to the `coinbaseWallet` connector when setting up wagmi, and it'd handle transaction conversions when appropriate.

Would this be possible? Or is there a better way of going about this?

**Comment by 0xmkh** [comment]
@Ncookiez Ideally we would want to have this paymaster logic to happen in Wagmi only because of maintenance reasons. Are you able to look for `coinbaseWallet` id from `connector?.[0]?.id` in `useAccount` hook and then add the paymaster logic with [useCapabilities](https://wagmi.sh/react/api/hooks/useCapabilities#usecapabilities) and [useWriteContracts](https://wagmi.sh/react/api/hooks/useWriteContracts#usewritecontracts) ? FYI there can be some unexpected behaviors with connectors, but we're working on fixing it ([here](https://github.com/wevm/wagmi/issues/3606) is the issue).

Let me know if this would work in your case 👍

**Reply by Ncookiez** [reply]
I definitely would need some logic to identify the smart wallet (connector id, chain id, etc.), but what I'm trying to avoid is duplicating and modifying all the transaction code on my app in order to support one wallet.

But I see your point with maintaining the connector - would this discussion be better placed on the wagmi repo? Or would this be something coinbase itself would have to consider supporting/developing on their end?

**Reply by 0xmkh** [reply]
Got it. We can try do something in RainbowKit. Can you list everything you'd like to have from RainbowKit and we can have a look ?

**Reply by Ncookiez** [reply]
I'd like to provide paymaster details and have the transaction conversions be handled by the underlying infra (aka not my app's code). Whether that's rainbowkit, wagmi, the coinbase wallet connector or even ethers I don't really mind, but I don't know what the best way to proceed from here would be.

**Reply by 0xmkh** [reply]
The best way to do it for now is via Wagmi. Otherwise we'll need more time to have such feature in RainbowKit.
